File: backend/app/models/investigator.py
from datetime import datetime

from sqlalchemy import Boolean, DateTime, String
from sqlalchemy.orm import Mapped, mapped_column

from app.db.session import Base
from app.models.mixins import TimestampMixin


class Investigator(TimestampMixin, Base):
    __tablename__ = "investigators"

    id: Mapped[int] = mapped_column(primary_key=True)
    investigator_id: Mapped[str] = mapped_column(String(30), unique=True, index=True)
    name: Mapped[str] = mapped_column(String(200))
    email: Mapped[str] = mapped_column(String(255), unique=True, index=True)
    role: Mapped[str] = mapped_column(String(40), index=True)
    role_label: Mapped[str] = mapped_column(String(80))
    department: Mapped[str] = mapped_column(String(160))
    clearance: Mapped[str] = mapped_column(String(120))
    status: Mapped[str] = mapped_column(String(30), default="ACTIVE", index=True)
    password_hash: Mapped[str] = mapped_column(String(255))
    can_view_all_cases: Mapped[bool] = mapped_column(Boolean, default=False)
    can_assign_cases: Mapped[bool] = mapped_column(Boolean, default=False)
    can_generate_reports: Mapped[bool] = mapped_column(Boolean, default=True)
    can_review_audit_trail: Mapped[bool] = mapped_column(Boolean, default=False)
    last_login: Mapped[datetime | None] = mapped_column(DateTime(timezone=True), nullable=True)

File: frontend/src/types/investigator.ts
export type InvestigatorRole = "INVESTIGATOR" | "SENIOR_INVESTIGATOR" | "ADMINISTRATOR";

export type RolePermissions = {
  canViewAllCases: boolean;
  canAssignCases: boolean;
  canGenerateReports: boolean;
  canReviewAuditTrail: boolean;
};

export type Investigator = {
  id: string;
  name: string;
  email: string;
  role: InvestigatorRole;
  role_label: string;
  department: string;
  clearance: string;
  status: "ACTIVE" | "SUSPENDED";
  last_login: string;
  permissions: RolePermissions;
};

export type AuditEvent = {
  id: string;
  investigator_id: string;
  investigator_name: string;
  action: string;
  target_type: "AUTH" | "CASE" | "ENTITY" | "EVIDENCE" | "REPORT" | "SYSTEM";
  target_id?: string;
  summary: string;
  created_at: string;
};

export type GeneratedReport = {
  id: string;
  case_id: number;
  case_number: string;
  title: string;
  investigator_id: string;
  created_at: string;
  content: string;
};
